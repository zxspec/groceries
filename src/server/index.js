const path = require("path");
const express = require("express");

const ENV_PATH = path.resolve(__dirname, "../../.env");
require("dotenv").config({ path: ENV_PATH });

const PORT = process.env.PORT || 9999;
const app = express();

app.get("*", (_, res) => {
  const content = "Hello from Groceries Search!";
  res.set("content-type", "text/html");
  res.send(content);
});

app.listen(PORT, () => console.log(`Server started http://localhost:${PORT}`));
