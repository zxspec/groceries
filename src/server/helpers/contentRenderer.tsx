import axios from "axios";
import React from "react";
import { Router } from "router5";
import { Provider } from "react-redux";
import { RouterProvider } from "react-router5";
import { renderToString } from "react-dom/server";

import App from "../../client/App";
import { createRootStore } from "../../client/store/createStore";

type Props = { router: Router };
export function render({ router }: Props) {
  // TODO migrate one level up
  const axiosInstance = axios.create({ baseURL: "/api" });
  const store = createRootStore({}, axiosInstance);

  const AppWithRouter = (
    <RouterProvider router={router}>
      <Provider store={store}>
        <App />
      </Provider>
    </RouterProvider>
  );

  const content = renderToString(AppWithRouter);

  return `
        <!DOCTYPE html>
        <html lang="en">
        <head>
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title>Grocery Search</title>
            <link rel="stylesheet" type="text/css" href="/main.css">
        </head>
        <body>
            <div id="react-root">${content}</div>
        </body>
        <script src="/bundle.js"></script>
        </html>
    `;
}
