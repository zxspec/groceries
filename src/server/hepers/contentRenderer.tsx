import React from "react";
import { Router } from "router5";
import { RouterProvider } from "react-router5";
import { renderToString } from "react-dom/server";

import App from "../../client/App";

type Props = { router: Router };
export default ({ router }: Props) => {
  const AppWithRouter = (
    <RouterProvider router={router}>
      <App />
    </RouterProvider>
  );

  const content = renderToString(AppWithRouter);

  return `
        <!DOCTYPE html>
        <html lang="en">
        <head>
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title>Document</title>
        </head>
        <body>
            <div id="react-root">${content}</div>
        </body>
        </html>
    `;
};
