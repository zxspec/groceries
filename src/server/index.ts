import path from "path";
import dotenv from "dotenv";
import express, { Request, Response } from "express";
import { cloneRouter } from "router5";
import axios from "axios";
import expressStaticGzip from "express-static-gzip";

import type { Grocery } from "../types";
import fakeGroceriesData from "./mock/groceries.json";
import { extractSearchPhrase, filterGroceriesByName } from "./api/filtering";
import createRouter from "../router/router";
import { render } from "./helpers/contentRenderer";
import { getStatusCode } from "./helpers/getStatusCode";
import { preloadData } from "./helpers/preloadData";
import { createRootStore } from "../client/store/createStore";

const ENV_PATH = path.resolve(__dirname, ".env");
console.debug("### ENV_PATH: ", ENV_PATH);
dotenv.config({ path: ENV_PATH });

const PORT = process.env.PORT || 8080;

const app = express();
const baseRouter = createRouter();
const axiosInstance = axios.create({ baseURL: `http://localhost:${PORT}/api` });

app.use(
  expressStaticGzip(path.join(__dirname, "../public"), {
    enableBrotli: true,
    orderPreference: ["br", "gz"],
  })
);

app.get("/api/search", (req: Request, res: Response) => {
  const groceries: Grocery[] = fakeGroceriesData; // or get it asynchronously from real API

  const searchPhrase = extractSearchPhrase(req.query);
  const data = filterGroceriesByName(groceries, searchPhrase);

  res.set("content-type", "application/json");
  res.send(data);
});

app.get("*", (req: Request, res: Response) => {
  const router = cloneRouter(baseRouter);

  router.start(req.originalUrl, async function done(error, route) {
    if (error) {
      res.status(500).send(error);
    } else {
      const store = createRootStore({}, axiosInstance);

      try {
        await preloadData(route, store.dispatch);
      } catch (e) {
        console.error("### preloadData error: ", e);
      }

      const content = render({ router, store });

      res.set("content-type", "text/html");
      res.status(getStatusCode(route.name)).send(content);
    }
  });
});

app.listen(PORT, () =>
  console.log(`### Server started http://localhost:${PORT}`)
);
