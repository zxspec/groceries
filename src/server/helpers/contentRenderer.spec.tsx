import { Router } from "router5";
import { renderToString } from "react-dom/server";
jest.mock("react-dom/server", () => ({
  renderToString: jest.fn(),
}));
import type { Store } from "redux";

import { render } from "./contentRenderer";

describe("contentRenderer", () => {
  const fakeRouter = {} as Router;
  const fakeStore = {} as Store;

  beforeAll(() => {
    fakeStore.subscribe = jest.fn();
    fakeStore.dispatch = jest.fn();
    fakeStore.getState = jest.fn();
  });

  afterEach(() => jest.clearAllMocks());

  it("should use `renderToString()` to render server-side content", () => {
    const result = render({ router: fakeRouter, store: fakeStore });

    expect(renderToString).toHaveBeenCalledTimes(1);
    expect(result).toMatchSnapshot();
  });
});
