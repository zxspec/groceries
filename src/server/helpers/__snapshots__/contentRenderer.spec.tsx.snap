// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contentRenderer should use \`renderToString()\` to render server-side content 1`] = `
"
        <!DOCTYPE html>
        <html lang=\\"en\\">
        <head>
            <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
            <title>Grocery Search</title>
            <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/main.css\\">
        </head>
        <body>
            <div id=\\"react-root\\">undefined</div>
        </body>
        <script>
          window.INITIAL_STATE=undefined
        </script>
        <script src=\\"/bundle.js\\"></script>
        </html>
    "
`;
